### DISCUSSION TITLE
[React] useListNavigation

### DISCUSSION THREAD

**Discussion by Marsony** [floating-ui/floating-ui#2234] [Q&A] [answered]
Hello everyone, thanks for this great library!

Is it possible to use 'useListNavigation' without the floating element?
For example, I want to make my Tabs and some vertical navigations accessible by keyboard.
So, useListNavigation can make it easily.

Ty!

**Comment by atomiks** [comment, marked as answer]
I've also thought about this and think it should be its own library that Floating UI wraps specifically for floating elements. I think other libraries have something similar called `Composite`. You can hack it to work already if you `useFloating` with `open: true` prop and pretend it's an always-open floating element without a reference anchor element.

**Reply by atomiks** [reply]
Demo: https://codesandbox.io/s/agitated-water-lifg5g?file=/src/App.tsx
I'll consider making it a standalone library before v1 of the lib is released =)

**Reply by Marsony** [reply]
> Demo: https://codesandbox.io/s/agitated-water-lifg5g?file=/src/App.tsx I'll consider making it a standalone library before v1 of the lib is released =)

Thanks! Floating-ui have a perfect versatility, with that package anyone can make anything

**Reply by Marsony** [reply]
> I've also thought about this and think it should be its own library that Floating UI wraps specifically for floating elements. I think other libraries have something similar called `Composite`. You can hack it to work already if you `useFloating` with `open: true` prop and pretend it's an always-open floating element without a reference anchor element.

One more question please..

We're trying to make our primitives like radix-ui, but we have a troubles with activeIndex.

In the example of the useListNavigation you're using options.map((a, **index**)). But in some cases we haven't access to the '**index**' because we're don't using the .map of our childs. (We're using only infinite nesting of the {children})

Can listRef accepts maybe Record<string, HTMLElement> ? Or maybe there is another way to solve the activeIndex trouble

**Reply by atomiks** [reply]
If you mean:

```js
<List>
  <Option />
  <Option />
  <Option />
</List>
```

`List` can render:

```js
React.Children.map(children, (child, index) => {
  // You can either use context or a prop with `cloneElement` 
  // to provide the index and `listRef` to `Option`. Then it will
  // know if it's active, and can add its element to `listRef`.
});
```

**Reply by Marsony** [reply]
> If you mean:
> 
> ```js
> <List>
>   <Option />
>   <Option />
>   <Option />
> </List>
> ```
> 
> `List` can render:
> 
> ```js
> React.Children.map(children, (child, index) => {
>   // You can either use context or a prop with `cloneElement` 
>   // to provide the index and `listRef` to `Option`. Then it will
>   // know if it's active, and can add its element to `listRef`.
> });
> ```

Sure, but also our Options can be wrapped in Group, GroupGroups, random divs and etc

**Reply by atomiks** [reply]
You can check `child.type === WhatComponent`. See: https://gist.github.com/danieljpgo/b3cca5c264da1ed243a1dd20aa34befe

Although, React itself is hinting that this Children API is "legacy" and suggests using JS data structures (I think JSX data structures looks more like native HTML personally): https://react.dev/reference/react/Children#alternatives

But yeah, feel free to open an issue if you think we can base it off a unique string instead of an index instead?

**Reply by Marsony** [reply]
> You can check `child.type === WhatComponent`. See: https://gist.github.com/danieljpgo/b3cca5c264da1ed243a1dd20aa34befe
> 
> Although, React itself is hinting that this Children API is "legacy" and suggests using JS data structures (I think JSX data structures looks more like native HTML personally): https://react.dev/reference/react/Children#alternatives
> 
> But yeah, feel free to open an issue if you think we can base it off a unique string instead of an index instead?

I thought maybe Record<useId(), HTMLElement> could help to resolve same cases. I'll open an issue.
But also I think it will make a troubles with any rerenders (strict mode)...
Thank you!
